include/revng/ADT/GenericGraph.h

```diff
@@ -388,6 +390,488 @@ class BidirectionalNode
   NeighborContainer Predecessors;
 };
 
+template<typename T>
+concept IsBidirectionalNode = requires {
+  T::is_bidirectional_node;
+  typename llvm::Inverse<T *>;
+};
+
+namespace detail {
+/// The parameters deciding specifics of the base type `NewNode` extends are
+/// non-trivial. That's why those decisiion were wrapped inside this struct
+/// to minimize clutter.
+///
+/// \note At this step `TheNode` has not been declared yet, thus we accept a
+///       template parameter that has the same signature as `TheNode` that
+///       will be declared later. This allows us to use it as if it was
+///       declared, provided that only the real `TheNode` is used as this
+///       argument.
+template<typename Node,
+         typename EdgeLabel,
+         bool HasParent,
+         size_t SmallSize,
+         template<typename, typename, bool, size_t, typename, size_t, bool>
+         class TheNode,
+         typename FinalType,
+         size_t ParentSmallSize,
+         bool ParentHasEntryNode>
+struct NewNodeBaseTCalc {
+  static constexpr bool
+    NoDerivation = std::is_same_v<FinalType, std::false_type>;
+  using NodeType = TheNode<Node,
+                           EdgeLabel,
+                           HasParent,
+                           SmallSize,
+                           FinalType,
+                           ParentSmallSize,
+                           ParentHasEntryNode>;
+  using DerivedType = std::conditional_t<NoDerivation, NodeType, FinalType>;
+  using GenericGraph = GenericGraph<DerivedType,
+                                    ParentSmallSize,
+                                    ParentHasEntryNode>;
+  using ParentType = Parent<GenericGraph, Node>;
+  using Result = std::conditional_t<HasParent, ParentType, Node>;
+};
+
+template<typename NodeType, typename LabelType>
+struct OwningEdge {
+  NodeType *Neighbor;
+  std::unique_ptr<LabelType> Label;
+};
+template<typename NodeType, typename LabelType>
+struct NonOwningEdge {
+  NodeType *Neighbor;
+  LabelType *Label;
+};
+template<typename NodeType, typename LabelType>
+struct EdgeView {
+  NodeType &Neighbor;
+  LabelType &Label;
+
+  explicit EdgeView(OwningEdge<NodeType, LabelType> &E) :
+    Neighbor(*E.Neighbor), Label(*E.Label) {}
+  explicit EdgeView(NonOwningEdge<NodeType, LabelType> &E) :
+    Neighbor(*E.Neighbor), Label(*E.Label) {}
+};
+template<typename NodeType, typename LabelType>
+struct ConstEdgeView {
+  NodeType const &Neighbor;
+  LabelType const &Label;
+
+  explicit ConstEdgeView(OwningEdge<NodeType, LabelType> const &E) :
+    Neighbor(*E.Neighbor), Label(*E.Label) {}
+  explicit ConstEdgeView(NonOwningEdge<NodeType, LabelType> const &E) :
+    Neighbor(*E.Neighbor), Label(*E.Label) {}
+};
```

Comment:
Drop in favor of `const EdgeView`?

Reply:
@aleclearmind Wait, I'm probably misunderstanding something, but I ran into problems trying to replace it. Normal `EdgeView` uses l-references. As a result, something like `[](OwningEdge const &E) -> EdgeView const { return EdgeView(E); };` can't be written, since `EdgeView` cannot be constructed from a const `OwningEdge ` object.

Did I do something wrong?

Reply:
Sad but we probably we can't do this, unless we `const_cast`.
But yeah, I'm persuaded that that `ConstEdgeView` has a reason to exists.